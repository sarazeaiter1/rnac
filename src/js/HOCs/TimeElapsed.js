import React from 'react';

import PropTypes from 'prop-types';

class TimeElapsed extends React.Component {
  state = {
    secondsElapsed: 0
  };
  static propTypes = {
    children: PropTypes.node,
    start: PropTypes.bool
  };
  static defaultProps = {
    start: false
  };

  componentWillMount() {
    if (this.props.start) {
      this.startCounting();
    }
  }
  componentWillUnmount() {
    if (this.timer) {
      clearInterval(this.timer);
    }
  }
  componentWillReceiveProps(nextProps) {
    if (!this.props.start && nextProps.start) {
      this.startCounting();
    } else if (!nextProps.start && this.props.start) {
      this.stopCounting();
    }
  }

  render() {
    const newFreshProps = {
      start: this.props.start,
      secondsElapsed: this.state.secondsElapsed
    };
    if (!this.props.children) return null;

    return typeof this.props.children === 'function'
      ? this.props.children(newFreshProps)
      : React.cloneElement(this.props.children, newFreshProps);
  }

  timer;
  isCountingSeconds = false;
  startCounting = () => {
    this.stopCounting();
    this.timer = setInterval(() => {
      this.setState({
        secondsElapsed: this.state.secondsElapsed + 1
      });
    }, 1000);
  };
  stopCounting = () => {
    if (this.timer) clearInterval(this.timer);
    this.setState({
      secondsElapsed: 0
    });
  };
}

export default TimeElapsed;
