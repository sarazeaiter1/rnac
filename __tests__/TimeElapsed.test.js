import React from 'react';
import { Text } from 'react-native';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import TimeElapsed from '../src/js/HOCs/TimeElapsed';

describe('TimeElapsed', () => {
  test('exports', () => {
    expect(TimeElapsed).toMatchSnapshot();
  });
  test('renders null with no children', () => {
    // const children = () => <View />;
    const tree = renderer.create(<TimeElapsed start={false} />);
    expect(JSON.stringify(tree)).toEqual('null');
  });

  test('renders and accepts children as callable function', () => {
    // eslint-disable-next-line react/prop-types
    const children = ({ secondsElapsed }) => {
      return (
        <Text>
          {JSON.stringify({
            secondsElapsed
          })}
        </Text>
      );
    };
    const tree = renderer.create(
      <TimeElapsed start={false}>
        {children}
      </TimeElapsed>
    );
    expect(tree).toMatchSnapshot();
  });

  test('renders and accepts children as React Class Component', () => {
    class CompChildren extends React.Component {
      render() {
        // eslint-disable-next-line react/prop-types
        const { secondsElapsed } = this.props;
        return (
          <Text>
            {JSON.stringify({
              secondsElapsed
            })}
          </Text>
        );
      }
    }

    const tree = renderer.create(
      <TimeElapsed start={false}>
        <CompChildren />
      </TimeElapsed>
    );
    expect(tree).toMatchSnapshot();
  });

  test('same ui tree if component or function', () => {
    // eslint-disable-next-line react/prop-types
    const children = ({ secondsElapsed }) => {
      return (
        <Text>
          {JSON.stringify({
            secondsElapsed
          })}
        </Text>
      );
    };

    class CompChildren extends React.Component {
      render() {
        // eslint-disable-next-line react/prop-types
        const { secondsElapsed } = this.props;
        return (
          <Text>
            {JSON.stringify({
              secondsElapsed
            })}
          </Text>
        );
      }
    }

    const withClass = renderer.create(
      <TimeElapsed start={false}>
        <CompChildren />
      </TimeElapsed>
    );

    const withFunction = renderer.create(
      <TimeElapsed start={false}>
        {children}
      </TimeElapsed>
    );

    expect(JSON.stringify(withClass)).toEqual(JSON.stringify(withFunction));
  });
  test('test componentWillMount', () => {
    // eslint-disable-next-line react/prop-types
    const children = ({ secondsElapsed }) => {
      return (
        <Text>
          {JSON.stringify({
            secondsElapsed
          })}
        </Text>
      );
    };
    const Comp = shallow(
      <TimeElapsed start={true}>
        <Text>
          {10}
        </Text>
      </TimeElapsed>
    );
    Comp.renderer._instance._instance.componentWillMount();
    expect(Comp).toMatchSnapshot();
  });
  test('test componentWillUnmount', () => {
    // eslint-disable-next-line react/prop-types
    const children = ({ secondsElapsed }) => {
      return (
        <Text>
          {JSON.stringify({
            secondsElapsed
          })}
        </Text>
      );
    };
    const Comp = shallow(
      <TimeElapsed start={true}>
        <Text>
          {10}
        </Text>
      </TimeElapsed>
    );
    Comp.renderer._instance._instance.timer = setInterval(() => {}, 1000);
    Comp.renderer._instance._instance.componentWillUnmount();
    expect(Comp).toMatchSnapshot();
  });
  test('test componentWillReceiveProps to start counting', () => {
    // eslint-disable-next-line react/prop-types
    const children = ({ secondsElapsed }) => {
      return (
        <Text>
          {JSON.stringify({
            secondsElapsed
          })}
        </Text>
      );
    };
    const Comp = shallow(
      <TimeElapsed start={true}>
        <Text>
          {10}
        </Text>
      </TimeElapsed>
    );
    Comp.renderer._instance._instance.props = { start: false };
    Comp.renderer._instance._instance.componentWillReceiveProps({
      start: true
    });
    expect(Comp).toMatchSnapshot();
  });
  test('test componentWillReceiveProps to stop counting', () => {
    // eslint-disable-next-line react/prop-types
    const children = ({ secondsElapsed }) => {
      return (
        <Text>
          {JSON.stringify({
            secondsElapsed
          })}
        </Text>
      );
    };
    const Comp = shallow(
      <TimeElapsed start={true}>
        <Text>
          {10}
        </Text>
      </TimeElapsed>
    );
    Comp.renderer._instance._instance.props = { start: true };
    Comp.renderer._instance._instance.componentWillReceiveProps({
      start: false
    });
    expect(Comp).toMatchSnapshot();
  });
});
