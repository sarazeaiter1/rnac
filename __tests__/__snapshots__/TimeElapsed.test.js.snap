// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeElapsed exports 1`] = `[Function]`;

exports[`TimeElapsed renders and accepts children as React Class Component 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  {"secondsElapsed":0}
</Text>
`;

exports[`TimeElapsed renders and accepts children as callable function 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  {"secondsElapsed":0}
</Text>
`;
